util/notification.py

```diff
@@ -0,0 +1,121 @@
+import json
+import logging
+
+from data.database import (
+    ExternalNotificationEvent,
+    RepositoryNotification,
+    TagNotificationSuccess,
+    db_for_update,
+    get_epoch_timestamp_ms,
+)
+from data.model import db_transaction, oci
+from data.model.user import get_namespace_by_user_id
+from data.registry_model import registry_model
+from notifications import spawn_notification
+
+logger = logging.getLogger(__name__)
+BATCH_SIZE = 10
+
+# Define a constant for the SKIP_LOCKED flag for testing purposes,
+# since we test with mysql 5.7 which does not support this flag.
+SKIP_LOCKED = True
+
+
+def fetch_active_notification(event, task_run_interval_ms=5 * 60 * 60 * 1000):
+    """
+    task_run_interval_ms specifies how long a task must wait before being ran again.
+    """
+    with db_transaction():
+        try:
+            # Fetch active notifications that match the event_name
+            query = (
+                RepositoryNotification.select(
+                    RepositoryNotification.id,
+                    RepositoryNotification.method,
+                    RepositoryNotification.repository,
+                    RepositoryNotification.event_config_json,
+                )
+                .where(
+                    RepositoryNotification.event == event.id,
+                    RepositoryNotification.number_of_failures < 3,
+                    (
+                        RepositoryNotification.last_ran_ms
+                        < get_epoch_timestamp_ms() - task_run_interval_ms
+                    )
+                    | (RepositoryNotification.last_ran_ms.is_null(True)),
```

Comment:
make sure `last_ran_ms` is indexed

Reply:
yes, added in index [here](https://github.com/quay/quay/pull/2894/files#diff-884d5b5805e4e80294281e5415112e55f65fe5c96a1471e04d190120b73d5d5fR25-R30)